tools/clang/lib/Sema/SemaHLSL.cpp

```diff
@@ -6398,18 +6416,39 @@ bool HLSLExternalSource::MatchArguments(
     const ArTypeObjectKind *pTT =
         g_LegalIntrinsicTemplates[pArgument->uLegalTemplates];
     if (AR_TOBJ_UNKNOWN != Template[i]) {
-      if ((AR_TOBJ_SCALAR == Template[i]) &&
-          (AR_TOBJ_VECTOR == *pTT || AR_TOBJ_MATRIX == *pTT)) {
-        Template[i] = *pTT;
-      } else {
+      // See if a perfect match overload is available
+      while (AR_TOBJ_UNKNOWN != *pTT && AR_TOBJ_NOCAST != *pTT) {
+        if (Template[i] == *pTT)
+          break;
+        pTT++;
+      }
+
+      if (AR_TOBJ_UNKNOWN == *pTT) {
+        // Perfect match failed and casts are allowed.
+        // Try splats and truncations to get a match.
+        pTT = g_LegalIntrinsicTemplates[pArgument->uLegalTemplates];
         while (AR_TOBJ_UNKNOWN != *pTT) {
-          if (Template[i] == *pTT)
+          if (AR_TOBJ_SCALAR == Template[i] &&
+              (AR_TOBJ_VECTOR == *pTT || AR_TOBJ_MATRIX == *pTT)) {
+            // If a scalar was passed in and the expected value was
+            // matrix/vector convert to the template type for a splat.
+            // Only applicable to VectorTT and MatrixTT,
+            // since the vec/mtx has to be first in the list.
+            Template[i] = *pTT;
             break;
+          } else if (AR_TOBJ_VECTOR == Template[i] && AR_TOBJ_SCALAR == *pTT) {
+            // If a vector was passed in and the expected value was scalar
+            // convert to the template type for a truncation.
+            // Only applicable to ScalarTT,
+            // since the scalar has to be first in the list.
+            Template[i] = AR_TOBJ_SCALAR;
+            break;
+          }
```

Comment:
This is the crux of the change. Instead of checking for splats first, we check for perfect matches first. If that fails due to running out of options, we'll try splats AND truncations (which, in spite of a comment suggesting otherwise, were not allowed here). However, if the ending enum is `NOCAST`, then we don't allow the splat or truncation and the match fails.

The reordering of the splat check won't change any previous behavior directly as it only kicked in when only a single template type was allowed in the first place. It was necessary for the new behavior allowing NOCAST though.

Reply:
I'm guessing truncation of matrices aren't allowed? I'd expect that to take place in the else if condition. Should there be some diagnostic if we encounter a matrix in this case?